java/dataset/src/main/java/org/apache/arrow/dataset/jni/NativeScanner.java

```diff
@@ -122,12 +124,19 @@ public ArrowRecordBatch next() {
 
   @Override
   public Schema schema() {
+    if (schema != null) {
+      return schema;
+    }
     readLock.lock();
     try {
+      if (schema != null) {
+        return schema;
+      }
```

Comment:
dup code on 132-134 vs 127-129?

Reply:
Sorry Yuan I missed the comment before merging.

The codes applies double-checked locking to balance between performance and atomicity. Although may look tedious as we don't usually have concurrent call to this method.

For example: `schema` can happen to be assigned by another thread before acquiring the lock. So we have to check again inside the lock.